eth/watchers/pricefeedwatcher.go

```diff
@@ -0,0 +1,176 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to get description: %w", err)
+	}
+
+	currencyFrom, currencyTo, err := parseCurrencies(description)
+	if err != nil {
+		return nil, err
+	}
+
+	w := &PriceFeedWatcher{
+		ctx:            ctx,
+		baseRetryDelay: priceUpdateBaseRetryDelay,
+		priceFeed:      priceFeed,
+		currencyBase:   currencyFrom,
+		currencyQuote:  currencyTo,
+		priceUpdated:   make(chan eth.PriceData, 1),
+	}
+
+	err = w.updatePrice()
+	if err != nil {
+		return nil, fmt.Errorf("failed to update price: %w", err)
+	}
+
+	go func() {
+		ctx, cancel := context.WithCancel(w.ctx)
+		defer cancel()
+		ticker := newTruncatedTicker(ctx, priceUpdatePeriod)
+		w.watch(ctx, ticker)
+	}()
+
+	return w, nil
+}
+
+// Currencies returns the base and quote currencies of the price feed.
+// i.e. base = CurrentPrice() * quote
+func (w *PriceFeedWatcher) Currencies() (base string, quote string) {
+	return w.currencyBase, w.currencyQuote
+}
+
+// Current returns the latest fetched price data.
+func (w *PriceFeedWatcher) Current() eth.PriceData {
+	return w.current
+}
+
+// PriceUpdated returns a channel that will receive the updated price data as
+// soon as it changes.
+func (w *PriceFeedWatcher) PriceUpdated() <-chan eth.PriceData {
+	return w.priceUpdated
+}
+
+func (w *PriceFeedWatcher) updatePrice() error {
+	newPrice, err := w.priceFeed.FetchPriceData()
+	if err != nil {
+		return fmt.Errorf("failed to fetch price data: %w", err)
+	}
+
+	if newPrice.UpdatedAt.After(w.current.UpdatedAt) {
+		w.current = newPrice
+		select {
+		case w.priceUpdated <- newPrice:
+		default:
+		}
+	}
+
+	return nil
+}
+
+func (w *PriceFeedWatcher) watch(ctx context.Context, ticker <-chan time.Time) {
+	for {
+		select {
+		case <-w.ctx.Done():
+			return
+		case <-ticker:
+			attempt, retryDelay := 1, w.baseRetryDelay
+			for {
+				err := w.updatePrice()
+				if err == nil {
+					break
+				} else if attempt >= priceUpdateMaxRetries {
+					clog.Errorf(ctx, "Failed to fetch updated price from PriceFeed attempts=%d err=%q", attempt, err)
+					break
+				}
+
+				clog.Warningf(ctx, "Failed to fetch updated price from PriceFeed, retrying after retryDelay=%d attempt=%d err=%q", retryDelay, attempt, err)
+				time.Sleep(retryDelay)
+				attempt, retryDelay = attempt+1, retryDelay*2
+			}
+		}
+	}
+}
+
+// parseCurrencies parses the base and quote currencies from a price feed based
+// on Chainlink PriceFeed description pattern "FROM / TO".
+func parseCurrencies(description string) (currencyBase string, currencyQuote string, err error) {
+	currencies := strings.Split(description, "/")
+	if len(currencies) != 2 {
+		return "", "", fmt.Errorf("aggregator description must be in the format 'FROM / TO' but got: %s", description)
+	}
+
+	currencyBase = strings.TrimSpace(currencies[0])
+	currencyQuote = strings.TrimSpace(currencies[1])
+	return
+}
+
+// newTruncatedTicker creates a ticker that ticks at the next time that is a
```

Comment:
1. Could you explain why the normal ticker couldn't be used for pricefeedwatcher? It's hard to understand why the standard ticker (or timer) would not work.

2. I'm also not sure we're not complicating things too much with this. I'm starting to understand what you want to achieve - that everyone B and O will update the price at the exactly same time. But I think it may be hard to achieve in practice because 1) the clocks are not in sync 2) There are network delays 3) Someone may start the in the middle of an hour. 

Maybe it'd be enough to update the price at the beginning of each round?

Reply:
1. It is in order to have _some_ synchronization across nodes, so it depends only the clock difference and network delays, but not the time when the node started. This would render a variance of a couple seconds at most, not up to an hour difference (if we use the same period). Even though we can never solve it perfectly, trying to reduce this desynchronization is worth it IMO.
2. As mentioned above, clock differences and network delays render a much lower desynchronization that if we did nothing at all. So my approach here was to propose something not too complex (the tests themselves were the only complex part IMO) that could get nodes a bit more synchronized.
 - Regarding the startup problem, we could build some logic to go around that, but that would seem like more complexity here (or for any other `watch` solution, like round-based) for a problem that I don't think is _that_ relevant. The node could be off-sync for at most an hour, which is already the worst-case scenario with this hourly-based approach anyway. So I'd rather not add further logic (like checking previous "price feed rounds" on startup) to go around that case, coming from the same mindset of not going too much out of our way to fix this "impossible" problem (the pricing desynchronization across nodes).

> Maybe it'd be enough to update the price at the beginning of each round?

IMO this would be more complex in the code (probably more LoC and requiring an integration with other bits of the codebase) and also increase the pricing risk by a lot. We could spend a whole day paying way more than we should, given ETH price volatility, while by updating every hour we get almost all data available that Chainlink oracles provide (I've seen them update roughly every hour), while also not being expensive in RPC calls etc. 

--

In conclusion, IMO this is an approach with a good trade-off between code complexity and accuracy. You've probably given more thought about this problem than I have tho, so let me know if you have a strong opinion towards reimplementing this to be round-based or just dropping the truncated ticker and using a regular one.